packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,282 @@
+import * as fs from 'fs-extra'
+import * as path from 'path'
+import { EventEmitter } from 'events'
+import chokidar, { FSWatcher } from 'chokidar'
+import { debounce } from 'mini-debounce'
+import { PagesData } from './pages'
+import {
+  extractStaticData,
+  defaultPageLoader,
+  defaultPageFinder,
+} from './utils'
+
+export class PageStrategy extends EventEmitter {
+  readonly getPages: () => Promise<Readonly<PagesData>>
+  readonly close: () => void
+
+  constructor(
+    pagesDir: string,
+    findPages: FindPages = defaultPageFinder,
+    loadPageData: LoadPageData = defaultPageLoader
+  ) {
+    super()
+
+    const pageCache: PagesData = {}
+    const fileCache: FileCache = {}
+    const watchers = new Set<FSWatcher>()
+
+    // The "promise" event is for HMR updates to the pages map.
+    const emitPromise = debounce(() => {
+      this.emit('promise', pagesPromise)
+    }, 100)
+
+    // The "change" event is for HMR updates to each page.
+    const changedPages = new Set<string>()
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const setPageData = (pageId: string, pageData: Partial<PageData>) => {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `setPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      const key = pageData.key ?? 'main'
+      if (pageData.dataPath) {
+        if (page.data[key])
+          throw Error(
+            `setPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = pageData.dataPath
+      }
+      if (pageData.staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `setPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = pageData.staticData
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    const removePageData = (pageId: string, key?: string) => {
```

Comment:
I prefer not providing a "delete" operation. I like React's philosophy: developers only need to decide how to map current state to a JSX tree. They don't need to decide when to delete a component, because React automatically take care of them. Similarly, users of vite-pages only need to decide how to map a `file path`(no matter it is added or changed) into `PageData`. Vite-pages can take care of the rest.

Reply:
Unless we know which file is associated with every `setPageData` call, we need to provide a way to manually remove pages. This is yet another reason to do what I suggested here: https://github.com/vitejs/vite-plugin-react-pages/pull/9#discussion_r587559865